metpy/calc/thermo.py

```diff
@@ -257,8 +264,47 @@ def dt(t, p):
                 / (mpconsts.Cp_d + (mpconsts.Lv * mpconsts.Lv * rs * mpconsts.epsilon
                                     / (mpconsts.Rd * t * t)))).to('kelvin')
         return frac / p
-    return units.Quantity(si.odeint(dt, atleast_1d(temperature).squeeze(),
-                                    pressure.squeeze()).T.squeeze(), temperature.units)
+
```

Comment:
Gist of my comments that got swallowed was that maybe this is a simpler logic tree and/or some combination? not fully tested

```python
    def dt(t, p):
        t = units.Quantity(t, temperature.units)
        p = units.Quantity(p, pressure.units)
        rs = saturation_mixing_ratio(p, t)
        frac = ((mpconsts.Rd * t + mpconsts.Lv * rs)
                / (mpconsts.Cp_d + (mpconsts.Lv * mpconsts.Lv * rs * mpconsts.epsilon
                                    / (mpconsts.Rd * t * t)))).to('kelvin')
        return frac / p

    if ref_pressure is None:
        ref_pressure = pressure[0]
    else:
        ref_pressure = ref_pressure.to('hPa')

    pressure = pressure.to('hPa')

    # Sort the pressures so they are in increasing order, save the order so we can return it
    sort_idx = np.argsort(pressure)
    pressure = pressure[sort_idx]

    # Find index of where reference pressure is (or would be)
    ref_pressure_idx = np.searchsorted(pressure.m, ref_pressure.m, side='left')
    inserted_ref_pressure = False
    if ref_pressure not in pressure:
        pressure = np.insert(pressure.m, ref_pressure_idx, ref_pressure.m) * pressure.units
        inserted_ref_pressure = True

    upward_slice = slice(ref_pressure_idx, None, None)
    if ref_pressure_idx == (len(pressure) - 1):
        downward_slice = slice(None, None, None)
    else:
        downward_slice = slice(None, ref_pressure_idx, None)

    # Integrate from reference pressure upward
    ret_temperatures = np.array([])
    if ref_pressure < pressure.max():
        ret_temperatures = np.append(ret_temperatures,
                                     si.odeint(dt, atleast_1d(temperature).squeeze(),
                                               pressure[upward_slice].squeeze()).T.squeeze())

    if ref_pressure > pressure.min():
        ret_temperatures = np.append(ret_temperatures,
                                     si.odeint(dt, atleast_1d(temperature).squeeze(),
                                               pressure[downward_slice].squeeze()).T.squeeze())

    # If we inserted the reference pressure, we need to remove it from pressure and the upward integral
    if inserted_ref_pressure:
        ret_temperatures = np.delete(ret_temperatures, ref_pressure_idx)
        pressure = np.delete(pressure.m, ref_pressure_idx)

    # Unsort things back to the user given order
    pressure = pressure[sort_idx]
    ret_temperatures = ret_temperatures[sort_idx]

    return units.Quantity(ret_temperatures.squeeze(), temperature.units)
```

Reply:
I updated `moist_lapse` to be a little clearer on the logic. I'm not wild about a few of the lines, but hopefully it's a little clearer what's going on. Let me know what you think.